packages/drivers/routerlicious-driver/src/documentService.ts

```diff
@@ -46,14 +46,15 @@ export class DocumentService implements api.IDocumentService {
             return new NullBlobStorageService();
         }
 
+        const storageToken = await this.tokenProvider.fetchStorageToken();
```

Comment:
What are the expectations here around failures / offline?
I.e. do we expect token provider to fail right away and rely on delta manager to do retries, or does token provider loops until it can get token? I do not remember now why, but in case of SPO/Bohemia, the choice was for token provider to hang in offline until it gets token.
Note that currently runtime only retries after container loaded. If any failure happens on snapshot loading, we immediately fail container load. This is mostly to allow host to have full control, as if we hang (while waiting for client to become online), then host has no object to check status - container was not yet returned to host in any form, so there is no way to cancel / stop any processes until it loads

Reply:
My current thinking is, socket token failure should be handled by connection and disconnection flow. For storage calls, we should keep retrying. Opened an issue to track that (#3960)

> Note that currently runtime only retries after container loaded. If any failure happens on snapshot loading, we immediately fail container load. This is mostly to allow host to have full control, as if we hang (while waiting for client to become online), then host has no object to check status - container was not yet returned to host in any form, so there is no way to cancel / stop any processes until it loads

Do you mean that the driver does not retry "getLatest" call failure and bubble up the error immediately to the host?

Reply:
Correct, driver always does one retry for 401/403 with new token (refresh = true), and otherwise has no other retry logic. Runtime will not retry getLatest (that is getSnapshotTree call in Container) - it will immediately fail.
I believe it would be great in the future (for this and other reasons) to return container object to caller right away, that way we can change these policies (if we want to) and client has something to close / kill or wait / report progress.

Reply:
BTW, same for storage calls - runtime handles retries for ops retrieval (see DeltaManager.getDeltas).
Currently we have no code to retry other operations (reading blobs & trees) - this has not been needed for SPO route as everything comes in one go (getLatest), but it's something we need now both for r11s and SPO routes.

Opened https://github.com/microsoft/FluidFramework/issues/3963 to track that.

Reply:
Correct for getLatest. I do not think retry logic should be in the driver - it's much better to have it in loader. Container can be closed at any moment and it would be great to stop all network activity. On top of that all the policies around how often to retry are right now in delta manager. I've opened an issue for that